### ISSUE TITLE
Add StyledText to specific range of text in GenericStyledArea<ParStyle, Either<String, LinkedImage>

### ISSUE THREAD

**Issue by anthonybazelle** [FXMisc/RichTextFX#740]
Hello,

First of all, thanks for your work!
I'm beginner with your API and JavaFX (I was on Swing so...) and I need it to set a background color to specific range of text in an area of type GenericStyledArea<ParStyle, Either<String, LinkedImage> like the RichTextDemo.java you provided.

I want to set specific CSS to this specific range of text. For example, is it possible to set a CSS to the world "Hello" in "Hello World !" ? Like a border-radius to the background color text, because I don't want a rectangle for the background color text. If it's not possible, is it possible with another component of your API like StyledTextArea, etc ?

Thanks !

**Comment by JordanMartinez** [comment]
> For example, is it possible to set a CSS to the world "Hello" in "Hello World !" ? Like a border-radius to the background color text

Yes, it's possible. A helpful resource for you to use would be the [CSS Reference Guide](https://github.com/FXMisc/RichTextFX/wiki/RichTextFX-CSS-Reference-Guide) if you don't already know about it.

> I need it to set a background color to specific range of text in an area of type `GenericStyledArea<ParStyle, Either<String, LinkedImage>` like the RichTextDemo.java you provided. I want to set specific CSS to this specific range of text. 

I'm not sure whether you want to know what CSS to use (in that case, see the reference guide linked above) or what code you need to write to make that happen (in that case, the RichTextDemo sort of provides such a thing).

Also, would you be interested in contributing to #624 by documenting the methods for Swing? It would save me time (as I could just fill in the corresponding part for RichTextFX). Also, I would know what's still not supported in this project.

**Comment by anthonybazelle** [comment]
Thanks for your reactivity and your answer ! 

Thanks for your CSS Reference guide, I didn't see it before. 
I suppose that something like code below will operate on a GenericStyledArea<> because inherit of StyledTextArea or should I change to a StyledTextArea ?
`.styled-text-area .text { 
-rtfx-underline-cap: <stroke-line-cap>; // Suppose I can use this to change the shape of underlined text, I will just looking for what properties I can use here.
}`

> I'm not sure whether you want to know what CSS to use (in that case, see the reference guide linked above) or what code you need to write to make that happen (in that case, the RichTextDemo sort of provides such a thing).

I think I need both, as I describe my problem in my first paragraph. Because I don't know if it's a good use to add more option like "set shape to underline text" in your TextStyle class in the RichTextDemo, or should I write a new class that refer to TextStyle class. Secondly, I need to set multiple underline color on different part of text, like the screenshot you provided but I need to do it only with code. If I change my property in the CSS to change the color of underline, will it change the color of all the previous underlined text or only the next ?

> Also, would you be interested in contributing to #624 by documenting the methods for Swing? It would save me time (as I could just fill in the corresponding part for RichTextFX). Also, I would know what's still not supported in this project.

I worked on Swing with Java 7, but only know your API with JavaFX. But if you want and when I have the time, I can look this function and document this.

**Comment by JordanMartinez** [comment]
Sorry, but I'm having a hard time understanding what you're saying at various points...

It sounds like you want to use an area to render both text and images to the screen. It also sounds like you want certain portions of the rendered text to have a background color that have a border around it. And it sounds like the issue is that the border part of the background doesn't have a `cap` CSS property. Is that correct?

**Comment by anthonybazelle** [comment]
> Sorry, but I'm having a hard time understanding what you're saying at various points...

Sorry my english is not very well

> It sounds like you want to use an area to render both text and images to the screen. It also sounds like you want certain portions of the rendered text to have a background color that have a border around it. And it sounds like the issue is that the border part of the background doesn't have a cap CSS property. Is that correct?

Yes it is. Basicly, I want to render several word that have different background color with a certain shape of background. For example I want something like this below but with different color not only purple (made with HTML/CSS) :

![image](https://user-images.githubusercontent.com/15831062/39618441-041485c2-4f84-11e8-8105-c0e91a1fc7af.png)

I think it will be more efficient with an example like this :)

**Comment by JordanMartinez** [comment]
Ok. That makes a lot more sense. You could try to add another `rtfx-` CSS property to TextExt to handle that, but I don't think that would be wise. Your use case seems rather unusual and it seems like that would take quite a bit of effort to get it right without screwing something else up along the way or without making maintenance on this project that much harder for me long-term.

The easiest way to implement this now is by using a custom `Segment` type. You'd need to render that segment type using a `Region` with two children: the background shape (which would allow you to do whatever you needed to make the shape you so desire) and the text that goes on top of it. The custom segment type itself would encapsulate all information needed to render such a thing (e.g. the String representing the text itself and any background shape-related info).

**Comment by JordanMartinez** [comment]
> I'm beginner in JavaFX, so I just need to understand your second solution. You propose to create a class that inherit from Segment and use Region ? And then in my GenericStyledArea<> or StyledTextArea I can use a list of this new class. Is that correct ?

`Segment` is a generic type, meaning you define what it is. Now, if you want to render both a regular Text object (e.g. `RegularText` which is just a `String`) and a custom text object with such a background (i.e. `CustomBackgroundText`), you usually use a functional-paradigm data structure known as `Either`. Think of Either as an abstract class with two concrete instances, `Left`, and `Right`. `Either<RegularText, CustomBackgroundText>` would be the Segment type you use, but you would pass in the concrete types of `Left<RegularText>` and `Right<CustomBackgroundText>`

Once you have defined the model part (the above Either), you then need to provide a way to render that model content onto the screen. For the `RegularText`, you'll just use a `TextExt`. For `CustomBackgroundText`, you'll need to use the approach I mentioned above`

Since you'll be using custom segment types, you'll need to use `GenericStyledArea`

**Comment by anthonybazelle** [comment]
Okay ! Thanks for your answer, I'll try that today and I'll come back

**Comment by JordanMartinez** [comment]
>  I can't use this in Scene Builder but I can with a StyleClassedTextArea.

I'm not sure. I don't use SceneBuilder. Perhaps its due to the usage of generics?

Also, since the custom background text would be wrapped in a `Region`, I'm not sure whether including it in a selection would work well (e.g. a selection made using `[` and `]` to denote the start and end of that selection in the text: `regular [text backgrou]nd text`)
